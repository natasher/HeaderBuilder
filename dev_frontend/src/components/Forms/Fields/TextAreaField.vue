<template>
  <div class = "row">

    <div class = "col col-left">
      <div class = "title">{{ fieldName }}</div>
    </div>

    <div class = "col col-right">
      <div class = "mfn-field">

        <textarea
          :name   = "fieldName"
          :id     = "wpId"
          v-model = "textAreaFieldValue">
        </textarea>

      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'

export default {

  name: 'TextAreaField',

  props: {
    /**
     * The name of the field.
     * Displayed on the left column.
     */
    fieldName: {
      type    : String,
      required: true,
    },
    /**
     * Name of the field used by WordPress internals.
     */
    wpId: {
      type   : String,
      require: true,
    },
  },

  methods: {
    ...mapActions( 'items', [ 'setCurrentFieldValue', ]),
  },

  computed: {
    ...mapGetters( 'fields', [ 'getCurrentFieldValue' ]),

    textAreaFieldValue: {
      get () {
        return this.getCurrentFieldValue( this.wpId )
      },
      set ( value ) {
        this.setCurrentFieldValue({
          name : this.wpId,
          value: value,
        })
      }
    }
  },
}
</script>