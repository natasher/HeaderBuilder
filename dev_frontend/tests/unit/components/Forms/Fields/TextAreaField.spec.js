import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import TextAreaField from '../../../../../src/components/Forms/Fields/TextAreaField.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

const itemVal = 'item form'

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      fields: {
        namespaced: true,
        getters: {
          getCurrentFieldValue: jest.fn().mockReturnValue(
            jest.fn().mockReturnValue( itemVal )
          ),
        },
      },
      items: {
        namespaced: true,
        actions: {
          setCurrentFieldValue: jest.fn()
        },
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    propsData: {
      fieldName: 'Test Field',
      wpId     : 'testField',
    }
  }

  return shallowMount( TextAreaField, merge( defaultMountingOptions, overrides ))
}

describe('TextAreaField.vue', () => {

  test('should match snapshot', () => {
    const wrapper = createWrapper()

    expect( wrapper.element ).toMatchSnapshot()
  })

  test('should get value from item form state', async () => {
    const innerGetter = jest.fn().mockReturnValue( itemVal )
    const getCurrentFieldValue = jest.fn().mockReturnValue( innerGetter )
    const store = createStore({
      modules: {
        fields: {
          getters: {
            getCurrentFieldValue,
          }
        }
      }
    })
    const wrapper = createWrapper({ store, })
    await flushPromises()

    expect(innerGetter).toBeCalledWith( wrapper.vm.wpId )
  })

  test('should set value if filled', async () => {
    const setCurrentFieldValue = jest.fn()
    const store = createStore({
      modules: {
        items: {
          actions: {
            setCurrentFieldValue,
          }
        }
      }
    })
    const wrapper = createWrapper({ store, })

    const textarea = wrapper.find( `textarea#${ wrapper.vm.wpId }` )
    textarea.trigger( 'input' )
    await flushPromises()

    expect(setCurrentFieldValue).toBeCalledWith(
      expect.anything(),
      {
        name : wrapper.vm.wpId,
        value: itemVal,
      },
      void 0,
    )
  })

})