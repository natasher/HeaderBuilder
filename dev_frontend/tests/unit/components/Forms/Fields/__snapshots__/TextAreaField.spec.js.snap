// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaField.vue should match snapshot 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <div
      class="title"
    >
      Test Field
    </div>
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field"
    >
      <textarea
        id="testField"
        name="Test Field"
      />
    </div>
  </div>
</div>
`;
